orttraining/orttraining/models/runner/training_runner.cc

```diff
@@ -279,62 +587,24 @@ Status TrainingRunner::TrainingLoop(IDataLoader& training_data_loader, IDataLoad
       // loop through the data
       size_t batch_num_cur_shard = training_data->TotalBatch(params_.batch_size);
       for (size_t batch = 0; batch < batch_num_cur_shard && step_ < params_.num_train_steps; ++batch) {
-        std::vector<MLValue> feeds = training_data->GetKthBatch(params_.batch_size, batch, input_allocator_);
-        if (loss_scaler_) {
-          float loss_scale = loss_scaler_->GetLossScale();
-          TrainingUtil::CreateCpuMLValue({1}, std::vector<float>{loss_scale}, &loss_scale_val, input_allocator_);
-          feeds.push_back(loss_scale_val);
-        }
-
-        {
-          float learning_rate = lr_scheduler->GetLearningRate(step_ + 1);
-          TrainingUtil::CreateCpuMLValue({1}, std::vector<float>{learning_rate}, &lr_ort_val, input_allocator_);
-          feeds.push_back(lr_ort_val);
-        }
+        const bool is_weight_update_step = (step_ + 1) % params_.gradient_accumulation_steps == 0;
 
+        VectorString feed_names;
+        VectorString fetch_names;
+        std::vector<MLValue> feeds;
         std::vector<MLValue> fetches;
 
-        const bool is_weight_update_step = (step_ + 1) % params_.gradient_accumulation_steps == 0;
-        auto start = std::chrono::high_resolution_clock::now();
+        PrepareFeedNamesAndFeeds(training_data_loader, *lr_scheduler, batch, feed_names, feeds);
```

Comment:
>PrepareFeedNamesAndFeeds [](start = 8, length = 24)

Could the preparation on names moved outside of loop? We should only need to attach Tensor inside loop, not dealing with names.

Reply:
It can be done outside the loop. But it largely reduce the readability and doesn't have much performance benefit. I personally would like to refactorize this runner into a more generic class, and this change is helpful to that  --- The logic is simplified from 
```
Do some names, do something else, add some other names, do something else, ..., (recall the names we added), add some feeds
```
to
```
Do some names and feeds
```
In the first case, data preparation spreads around, while I feel data preparation deserves an isolated and clean space as shown in the second case.

Reply:
OK let's do the refactor later.

---
In reply to: [419606588](https://github.com/microsoft/onnxruntime/pull/3770#discussion_r419606588) [](ancestors = 419606588)